src/kernels/MIOpenReduceCalculation.cpp

```diff
@@ -53,71 +59,140 @@ __device__ void calculationparallelfwdcontiguous(const TI* __restrict__ x,
+                                                            const uint64_t inner_size,
+                                                            const bool nanPropagation)
 {
     // instantiate the kernel
-    calculationparallelfwdcontiguous<INPUT_TYPE, OUTPUT_TYPE, OP_TYPE>(
+    calculationparallelfwdcontiguous<INPUT_TYPE,
+                                     OUTPUT_TYPE,
+                                     OP_TYPE,
+                                     CALCULATION_DTYPE,
+                                     IS_LOGICAL_CALCULATION,
+                                     IS_USE_FLOAT_DTYPE>(
         x, y, output_numel, reduce_size, parallelism_size, inner_size, nanPropagation);
 }
 
-template <typename TI, typename TO, ReduceCalculationOp_t op>
+template <typename TI,
+          typename TO,
+          ReduceCalculationOp_t op,
+          typename CAL_TYPE,
+          bool IS_LOGICAL,
+          bool IS_USE_FLOAT>
 __device__ void calculationfwdcontiguous(const TI* __restrict__ x,
                                          TO* __restrict__ y,
-                                         uint64_t output_numel,
-                                         uint64_t reduce_size,
-                                         uint64_t inner_size,
-                                         bool nanPropagation)
+                                         const uint64_t output_numel,
+                                         const uint64_t reduce_size,
+                                         const uint64_t inner_size,
+                                         const bool nanPropagation)
 {
     const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
     if(gid >= output_numel)
         return;
 
     uint64_t input_idx = (gid / inner_size) * inner_size * reduce_size + gid % inner_size;
 
-    FLOAT_ACCUM calculation = static_cast<FLOAT_ACCUM>(0);
+    CAL_TYPE calculation = reduce_func<CAL_TYPE, op>{}.get_initial_value();
+
     for(uint64_t k = 0; k < reduce_size; ++k)
     {
-        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(x[input_idx]);
-        if(nanPropagation && isnan(val))
+        // preprocess val
+        CAL_TYPE val;
+        if constexpr(IS_LOGICAL || !IS_USE_FLOAT)
         {
-            val = static_cast<FLOAT_ACCUM>(0);
+            // For logical calculation or integer calculation, no need to convert to float
+            // because there's no need for precision
+            val = x[input_idx];
         }
-        reduce_func<FLOAT_ACCUM, op>{}.calculate(calculation, val);
+        else
+        {
+            val = CVT_FLOAT2ACCUM(x[input_idx]);
+        }
+
+        // handle nan
+        if constexpr(IS_USE_FLOAT && !IS_LOGICAL)
+        {
+            // Only need to check nan values for inputs with float type
+            // and for numerical calculation
+            if(nanPropagation && isnan(val))
+            {
+                val = static_cast<TI>(0);
```

Comment:
fix: for sum, nan should be converted to 0, but for prod, nan should be converted to 1